racket/collects/racket/private/choose-file-to-load.rkt

```diff
@@ -0,0 +1,177 @@
+                                            "native"
+                                            (system-library-subpath)
+                                            (if rep-sfx?
+                                                (path-add-extension
+                                                 file
+                                                 dll-suffix)
+                                                file)))))]
+               [zo (lambda (root-dir compiled-dir)
+                     (build-path (reroot base root-dir)
+                                 compiled-dir
+                                 (path-add-extension file #".zo")))]
+               [alt-zo (lambda (root-dir compiled-dir)
+                         (build-path (reroot base root-dir)
+                                     compiled-dir
+                                     (path-add-extension alt-file #".zo")))]
+               [so (get-so file #t)]
+               [alt-so (get-so alt-file #t)]
+               [try-main? (or main-path-d (not alt-path-d))]
+               [try-alt? (and alt-file (or alt-path-d (not main-path-d)))])
+          (cond
+            [(and allow-zo-so?
+                  so
+                  try-main?
+                  (date>=? modes roots so path-d))
+             => (lambda (so-d)
+                  (values #f 'so (car so-d)))]
+            [(and allow-zo-so?
+                  alt-so
+                  try-alt?
+                  (date>=? modes roots alt-so alt-path-d))
+             => (lambda (so-d)
+                  (values alt-path 'so (car so-d)))]
+            [(and allow-zo-so?
+                  try-main?
+                  (date>=? modes roots zo path-d))
+             => (lambda (zo-d)
+                  (values #f 'zo (car zo-d)))]
+            [(and allow-zo-so?
+                  try-alt?
+                  (date>=? modes roots alt-zo path-d))
+             => (lambda (zo-d)
+                  (values alt-path 'zo (car zo-d)))]
+            [(or (not (pair? expect-module))
+                 (car expect-module)
+                 (is-compiled-file? (if try-main? path alt-path)))
+             (let ([p (if try-main? path alt-path)])
+               (values (and expect-module
+                            (not try-main?)
+                            p) ;; current-module-declare-source
+                       'src ;; plain file
+                       (and (or (not (pair? expect-module))
+                                (file-exists? p)) ;; used to be `quiet-failure?`
+                            p)))]
+            [else
+             ;; here the load-handler should do nothing; this was a fallthru case in the other
+             (values #f 'src #f)]))))))
+
+(define-values (dll-suffix)
+  (system-type 'so-suffix))
+
+(define (is-compiled-file? p)
+  (and (file-exists? p)
+       (call-with-input-file* p
+         (λ (port)
+           (starts-with-compiled-object-bytes? port)))))
+
+(define (starts-with-compiled-object-bytes? p)
```

Comment:
Could this be the function "load-handler.rkt" and used from there, so there's not two (mostly) copies? There may be some obstacle to sharing that I'm not seeing.

Reply:
I tried to look into this a bit more and I'm not sure what's going on, actually. It looks like `linklet-directory-start` is exported from load-handler.rkt, but I can't find where it is used outside that file. If it isn't used, then I think it should be fine to move a bunch more code from load-handler.rkt to (the new) `racket/private/choose-file-to-load`. 

That said, I think the duplication is pretty small and the test case gives us above average coverage and confidence in `starts-with-compiled-object-bytes?`.

Reply:
I think it was used only by "handler.rkt". While it still seems better to me to restore the few lines in `racket/private/choose-file-to-load` to make it `linklet-bundle-or-directory-start` and share that — less about bugs and more about changing fewer places when the format inevitably changes — I'm ok with this organization.

Reply:
Okay, I see it now. It really isn't as much code as I thought it was going to be. (I'm imagining that these commits will be squashed into one before merging to master, just in case this is the last comment. I'm happy to do it, of course.)